File: tags/version-1.0/file_db/src/edu/ur/file/mime/TopMediaType.java
/**
  Copyright 2008 University of Rochester

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/
package edu.ur.file.mime;

import java.util.Collections;
import java.util.Set;
import java.util.HashSet;

import edu.ur.persistent.CommonPersistent;
import edu.ur.exception.DuplicateNameException;
/**
 * This represents a Mime top Media type
 * 
 * @author Nathan Sarr
 *
 */
public class TopMediaType extends CommonPersistent
{
	/** Unique serial id */
	private static final long serialVersionUID = -8615770105531118951L;
	
	/**  Set of Mime sub types for this top level type */
    private Set<SubType> subTypes = new HashSet<SubType>();
    
	
	/**
	 * Default Constructor 
	 */
	TopMediaType(){}
	
	
	/**
	 * Create a top media type with the specified name.
	 * 
	 * @param name
	 */
	public TopMediaType(String name)
	{
		setName(name);
	}
	
	/**
	 * Create a new top media type with the given name and description.
	 * 
	 * @param name
	 * @param description
	 */
	public TopMediaType(String name, String description)
	{
		this(name);
		setDescription(description);
	}
	
	/**
	 * Create a subType and return it.
	 * 
	 * @param name - name of the sub type
	 * @return
	 * 
	 * @throws DuplicateNameException if sub type with name already exists
	 */
	public SubType createSubType(String name) throws DuplicateNameException
	{
		if( this.getSubType(name) != null )
		{
			throw new DuplicateNameException("Top media type " + this + " already contains sub type", name );
		}
		SubType subType = new SubType(this, name);
		subTypes.add(subType);
		return subType;
	}
	
	/**
	 * Remove a subType from this subtype
	 * 
	 * @param subType to remove
	 * @return true if the subType is removed
	 */
	private boolean removeSubType(SubType subType)
	{
		return subTypes.remove(subType);
	}
	
	/**
	 * Remove a subtype
	 * 
	 * @param sub type name to remove
	 * @return true if the subtype is removed
	 */
	public boolean removeSubType(String subTypeName)
	{
		return removeSubType(getSubType(subTypeName));
	}
	
	/**
	 * Remove an extension from this subtype
	 * 
	 * @param id of extension to remove
	 * @return true if the extension is removed
	 */
	public boolean removeSubType(Long id)
	{
		return removeSubType(getSubType(id));
	}
	
	/**
	 * Returns the sub type if the sub type name exists otherwise 
	 * returns null
	 * 
	 * @param subTypeName
	 * @return the found sub type or null if it is not found
	 */
	public SubType getSubType(String subTypeName)
	{
		SubType subType = null;
		
		for( SubType type: subTypes)
		{
			if( type.getName().equals(subTypeName)){
				subType = type;
			}
		}
		
		return subType;
	}
	
	/**
	 * Returns the sub type if the id exists otherwise 
	 * returns null - null is always returned if the id is
	 * 0
	 * 
	 * @param Sub type id
	 * @return the found subType or null if it is not found
	 */
	public SubType getSubType(Long id)
	{
		SubType subType = null;
		
		if( id == 0 )
		{
			return subType;
		}
		
		for( SubType type: subTypes)
		{
			if( type.getId().equals(id)){
				subType = type;
			}
		}
		
		return subType;
	}

	/**
	 * The set of mime sub types for this top 
	 * level mime type
	 * 
	 * @return
	 */
	public Set<SubType> getSubTypes() {
		return Collections.unmodifiableSet(subTypes);
	}

	/**
	 * Set the set of mime sub types for this top level
	 * type
	 * 
	 * @param subTypes
	 */
	public void setSubTypes(Set<SubType> subTypes) {
		this.subTypes = subTypes;
	}
	
	/* (non-Javadoc)
	 * @see java.lang.Object#toString()
	 */
	public String toString()
	{
		
		StringBuffer sb = new StringBuffer("[TopMediaType name = ");
		sb.append(name);
		sb.append(" id = " );
		sb.append(id);
		sb.append( "description = ");
		sb.append(description);
		sb.append("]");
		return sb.toString();
	}
	
	/* (non-Javadoc)
	 * @see java.lang.Object#hashCode()
	 */
	public int hashCode()
	{
		return name == null ? 0 : name.hashCode();
	}
	
	/* (non-Javadoc)
	 * @see java.lang.Object#equals(java.lang.Object)
	 */
	public boolean equals(Object o)
	{
		if (this == o) return true;
		if (!(o instanceof TopMediaType)) return false;

		final TopMediaType topMediaType = (TopMediaType) o;

		if( ( name != null && !name.equals(topMediaType.getName()) ) ||
			( name == null && topMediaType.getName() != null ) ) return false;

		return true;
	}
}

File: tags/ir_plus-2.0.7/ir_web/webapps/ir_web/page-resources/js/admin/top_media_type.js
/*
   Copyright 2008 University of Rochester

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

/**
 * This code is for dealing with adding and removing folders 
 * in the workspace.
 */
YAHOO.namespace("ur.topMediaType");

// action to perform when submitting the personal content types.
var myTopMediaTypeAction = basePath + 'admin/getTopMediaTypes.action';

// actions for adding and removing top media types
var updateTopMediaTypeAction = basePath + 'admin/updateTopMediaType.action';
var newTopMediaTypeAction = basePath + 'admin/createTopMediaType.action';
var deleteTopMediaTypeAction = basePath + 'admin/deleteTopMediaType.action';
var getTopMediaTypeAction = basePath + 'admin/getTopMediaType.action';

// object to hold the specified content type data.
var myTopMediaTypesTable = new YAHOO.ur.table.Table('myTopMediaTypes', 'newTopMediaTypes');


YAHOO.ur.topMediaType = {

    /**
     *  Function that retireves content information
     *  based on the given content id.
     *
     *  The top media type id used to get the folder.
     */
    getTopMediaTypes : function(rowStart, startPageNumber, currentPageNumber, order)
    {
        /*
         * This call back updates the html when a new content type is
         * retrieved.
         */
        var callback =
        {
            success: function(o) 
            {
                // check for the timeout - forward user to login page if timout
	            // occured
	            if( !urUtil.checkTimeOut(o.responseText) )
	            {               
                    var divToUpdate = document.getElementById('newTopMediaTypes');
                    divToUpdate.innerHTML = o.responseText; 
                }
            },
	
	        failure: function(o) 
	        {
	            alert('Get top media type Failure ' + o.status + ' status text ' + o.statusText );
	        }
        };
        
        var transaction = YAHOO.util.Connect.asyncRequest('GET', 
            myTopMediaTypeAction + '?rowStart=' + rowStart 
	        					+ '&startPageNumber=' + startPageNumber 
	        					+ '&currentPageNumber=' + currentPageNumber 
	        					+ '&sortType=' + order 
	        					+ '&bustcache='+new Date().getTime(), 
            callback, null);
    }, 
    
    /**
     * Clear the top media type form
     */
    clearTopMediaTypeForm : function()
    {
        var div = document.getElementById('topMediaTypeError');
        div.innerHTML = "";
	    document.getElementById('newTopMediaTypeForm_name').value = "";
	    document.getElementById('newTopMediaTypeForm_description').value = "";
	    document.getElementById('newTopMediaTypeForm_id').value = "";
	    document.newTopMediaTypeForm.newTopMediaType.value = "true";
    },
    
    /**
     * Sets all checkboxes to checked
     */
    setCheckboxes : function()
    {
        checked = document.myTopMediaTypes.checkAllSetting.checked;
        var topMediaTypeIds = document.getElementsByName('topMediaTypeIds');
        urUtil.setCheckboxes(topMediaTypeIds, checked);
    },
    
    /**
     * Creates a new media type dialog
     */
    createNewTopMediaTypeDialog : function()
    {
    	// Define various event handlers for Dialog
	    var handleSubmit = function() 
	    {
	       this.submit();
	    };
	
	    // handle a cancel of the adding top media type dialog
	    var handleCancel = function() 
	    {
	        YAHOO.ur.topMediaType.clearTopMediaTypeForm();
	        YAHOO.ur.topMediaType.newTopMediaTypeDialog.hide();
	    };
	
	    var handleSuccess = function(o) 
	    {
	    	// check for the timeout - forward user to login page if timout
	        // occured
	        if( !urUtil.checkTimeOut(o.responseText) )
	        {               
	    	    //get the response from adding a content type
	            var response = o.responseText;
	            var topMediaTypeForm = document.getElementById('newTopMediaTypeDialogFields');
	    
	            // update the form fields with the response.  This updates
	            // the form, if there was an issue, update the form with
	            // the error messages.
	            topMediaTypeForm.innerHTML = o.responseText;
	    
	            // determine if the add/edit was a success
	            var success = document.getElementById("newTopMediaTypeForm_success").value;
	    
	            //if the top media type was not added then show the user the error message.
	            // received from the server
	            if( success == "false" )
	            {
                    YAHOO.ur.topMediaType.newTopMediaTypeDialog.showDialog();
	            }
	            else
	            {
	                // we can clear the form if the top media type was added
	                YAHOO.ur.topMediaType.newTopMediaTypeDialog.hide();
	                YAHOO.ur.topMediaType.clearTopMediaTypeForm();
	            }
	            myTopMediaTypesTable.submitForm(myTopMediaTypeAction);
	        }
	    };
	
	    // handle form sbumission failure
	    var handleFailure = function(o) 
	    {
	        alert('top media type submission failed ' + o.status + ' status text ' + o.statusText);
	    };

	    // Instantiate the Dialog
	    // make it modal - 
	    // it should not start out as visible - it should not be shown until 
	    // new top media type button is clicked.
	    YAHOO.ur.topMediaType.newTopMediaTypeDialog = new YAHOO.widget.Dialog('newTopMediaTypeDialog', 
            { width : "600px",
		      visible : false, 
		      modal : true,
		      buttons : [ { text:'Submit', handler:handleSubmit, isDefault:true },
					      { text:'Cancel', handler:handleCancel } ]
		    } );
		
		// override the submit function
        YAHOO.ur.topMediaType.newTopMediaTypeDialog.submit = function()
        {
            YAHOO.util.Connect.setForm('newTopMediaTypeForm');
	        if( YAHOO.ur.topMediaType.newTopMediaTypeDialog.validate() )
	        {
	            //based on what we need to do (update or create a 
	            // new top media type) based on the action.
                var action = newTopMediaTypeAction;
	            if( document.newTopMediaTypeForm.newTopMediaType.value != 'true')
	            {
	               action = updateTopMediaTypeAction;
	            }

                var cObj = YAHOO.util.Connect.asyncRequest('post', action, callback);
             }
        }
        
 	    // Validate the entries in the form to require that both first and last name are entered
	    YAHOO.ur.topMediaType.newTopMediaTypeDialog.validate = function() 
	    {
	        var name = document.getElementById('newTopMediaTypeForm_name').value;
		    if (name == "" || name == null) 
		    {
		        alert('A top media type name must be entered');
			    return false;
		    }     
		    else 
		    {
			    return true;
		    }
	    };
	    
	    //center and show the dialog
	    YAHOO.ur.topMediaType.newTopMediaTypeDialog.showDialog = function()
	    {
	        YAHOO.ur.topMediaType.newTopMediaTypeDialog.center();
	        YAHOO.ur.topMediaType.newTopMediaTypeDialog.show();
	    };

	    // Wire up the success and failure handlers
	    var callback = { success: handleSuccess, failure: handleFailure };
			
			
	    // Render the Dialog
	    YAHOO.ur.topMediaType.newTopMediaTypeDialog.render();

        // listener for showing the dialog when clicked.
	    YAHOO.util.Event.addListener("showTopMediaType", "click", 
	        YAHOO.ur.topMediaType.newTopMediaTypeDialog.showDialog, 
	        YAHOO.ur.topMediaType.newTopMediaTypeDialog, true);
	    
    },
    
    /**
     * function to edit top media type information
     */
    editTopMediaType : function(id)
    {	    
	    /*
         * This call back updates the html when a editing a mime top media type
         */
        var callback =
        {
            success: function(o) 
            {
            	// check for the timeout - forward user to login page if timout
	            // occured
	            if( !urUtil.checkTimeOut(o.responseText) )
	            {               
                    var divToUpdate = document.getElementById('newTopMediaTypeDialogFields');
                    divToUpdate.innerHTML = o.responseText; 
                    document.newTopMediaTypeForm.newTopMediaType.value = "false";
                    YAHOO.ur.topMediaType.newTopMediaTypeDialog.showDialog();
                }
            },
	
	        failure: function(o) 
	        {
	            alert('Edit top media type Failure ' + o.status + ' status text ' + o.statusText );
	        }
        };
        
        var transaction = YAHOO.util.Connect.asyncRequest('GET', 
            getTopMediaTypeAction + '?id=' + id +  '&bustcache='+new Date().getTime(), 
            callback, null);
    },
    
    /**
     * Clear the delete form
     */
    clearDeleteTopMediaTypeForm : function()
    {
         var div = document.getElementById('topMediaTypeError');
         div.innerHTML = "";
    },
    
    /**
     * Create the delete dialog
     */
    createDeleteTopMediaTypeDialog : function()
    {
    	// Define various event handlers for Dialog
	    var handleSubmit = function()
	    {
	        YAHOO.util.Connect.setForm('myTopMediaTypes');
	    
	        //delete the top media type
            var cObj = YAHOO.util.Connect.asyncRequest('post',
            deleteTopMediaTypeAction, callback);
	    };
	
		
	    // handle a cancel of deleting top media type dialog
	    var handleCancel = function() 
	    {
	        YAHOO.ur.topMediaType.deleteTopMediaTypeDialog.hide();
	    };
	
	    var handleSuccess = function(o) 
	    {
	    	// check for the timeout - forward user to login page if timout
	        // occured
	        if( !urUtil.checkTimeOut(o.responseText) )
	        {               
	            //get the response from adding a top media type
	            var response = eval("("+o.responseText+")");
	    
	            //if the top media type was not deleted then show the user the error message.
	            // received from the server
	            if( response.topMediaTypeDeleted == "false" )
	            {
	                var deleteTopMediaTypeError = document.getElementById('form_deleteTopMediaTypeError');
                    deleteTopMediaTypeError.innerHTML = '<p id="newDeleteTopMediaTypeError">' 
                    + response.message + '</p>';
                    YAHOO.ur.topMediaType.deleteTopMediaTypeDialog.showDialog();
	            }
	            else
	            {
	                // we can clear the form if the top media types were deleted
	                YAHOO.ur.topMediaType.clearDeleteTopMediaTypeForm();
	                YAHOO.ur.topMediaType.deleteTopMediaTypeDialog.hide();
	            }
	            // reload the table
	            myTopMediaTypesTable.submitForm(myTopMediaTypeAction);
	        }
	    };
	
	    // handle form submission failure
	    var handleFailure = function(o) 
	    {
	        alert('delete top media type failed ' + o.status + ' status text ' + o.statusText);
	    };

	    // Instantiate the Dialog
	    // make it modal - 
	    // it should not start out as visible - it should not be shown until 
	    // new top media type button is clicked.
	    YAHOO.ur.topMediaType.deleteTopMediaTypeDialog = new YAHOO.widget.Dialog('deleteTopMediaTypeDialog', 
            { width : "500px",
		      visible : false, 
		      modal : true,
		      buttons : [ { text:'Yes', handler:handleSubmit, isDefault:true },
					      { text:'No', handler:handleCancel } ]
		    } );
		
		YAHOO.ur.topMediaType.deleteTopMediaTypeDialog.showDialog = function()
		{
		    YAHOO.ur.topMediaType.deleteTopMediaTypeDialog.center();
		    YAHOO.ur.topMediaType.deleteTopMediaTypeDialog.show();
		}
   
	    // Wire up the success and failure handlers
	    var callback = { success: handleSuccess, failure: handleFailure };
			
	    // Render the Dialog
	    YAHOO.ur.topMediaType.deleteTopMediaTypeDialog.render();

        // listener for showing the dialog when clicked.
	    YAHOO.util.Event.addListener("showDeleteTopMediaType", "click", 
	        YAHOO.ur.topMediaType.deleteTopMediaTypeDialog.showDialog, 
	        YAHOO.ur.topMediaType.deleteTopMediaTypeDialog, true);
    },
    
    init : function()
    {
        YAHOO.ur.topMediaType.getTopMediaTypes(0,1,1,'asc');
        YAHOO.ur.topMediaType.createNewTopMediaTypeDialog();
        YAHOO.ur.topMediaType.createDeleteTopMediaTypeDialog();
    }   
}


// initialize the code once the dom is ready
YAHOO.util.Event.onDOMReady(YAHOO.ur.topMediaType.init);

